[DOC FIX] Show diagrams and elaborate on interaction between the client, the tracking server and stores (#4014)
mlflow/mlflow: docs/source/tracking.rst

MLflow entities (runs, parameters, metrics, tags, notes, metadata, etc), the artifact store persists artifacts
(files, models, images, in-memory objects, or model summary, etc).

The MLflow client can interface with a variety of `backend <https://mlflow.org/docs/latest/tracking.html#backend-stores>`_ and `artifact <https://mlflow.org/docs/latest/tracking.html#artifact-stores>`_ storage configurations.
Here are four common configuration scenarios:

Scenario 1: MLflow on localhost
-------------------------------

Many developers run MLflow on their local machine, where both the backend and artifact store share a directory
on the local filesystem—``./mlruns``—as shown in the diagram. The MLflow client directly interfaces with an
instance of a `FileStore` and `LocalArtifactRepository`.

.. figure:: _static/images/scenario_1.png

In this simple scenario, the MLflow client uses the following interfaces to record MLflow entities and artifacts:

 * An instance of a `LocalArtifactRepository` (to store artifacts)

>  * An instance of a `FileStore` (to save MLflow entities)

Reviewer: Can we add github links to `LocalArtifactRepository`, `FileStore`, etc (`S3ArtifactRepository` down below)?

Author: You read my mind. I was wondering if should. By down below, where did you mean? Should we include them as links
while we mention in each scenario, or should we add a paragraph at the end as such:

The [FileStore](https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/file_store.py#L115), [RestoreStore](https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/rest_store.py#L39), and [SQLAlchemyStore](https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/sqlalchemy_store.py#L61) are concrete implementations of the abstract class [AbstractStrore](https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/abstract_store.py), whereas the  [LocalArtifactRepository](https://github.com/mlflow/mlflow/blob/master/mlflow/store/artifact/local_artifact_repo.py#L15) and [S3ArtifactRepository](https://github.com/mlflow/mlflow/blob/master/mlflow/store/artifact/s3_artifact_repo.py#L14) are concrete implemetnations of the abstract class [ArtifactRepository](https://github.com/mlflow/mlflow/blob/master/mlflow/store/artifact/artifact_repo.py#L13)

Author: Added this after the note. Better place.

Scenario 2: MLflow on localhost with SQLite
-------------------------------------------

> Many users also run MLflow on their local machines with a `SQLAlchemy-compatible <https://docs.sqlalchemy.org/en/14/core/engines.html#database-urls>`_ database: **SQLite**. In this case, artifacts

Reviewer: Can we add the SQLite link here? https://sqlite.org/docs.html (I see we added it below - probably best to add it on the first instance)

Author: Okay, I will do it!

are stored under the local ``./mlruns`` directory, and MLflow entities are inserted in a `SQLite <https://sqlite.org/docs.html>`_ database file ``mlruns.db``.

.. figure:: _static/images/scenario_2.png

In this scenario, the MLflow client uses the following interfaces to record MLflow entities and artifacts:

>  * An instance of a `LocalArtifactRepository` (to save artifacts)

Reviewer: ```suggestion
 * An instance of a `LocalArtifactRepository` (to save artifacts to a local `mlruns` directory)
```

Author: done!

 * An instance of an `SQLAlchemyStore` (to save MLflow entities to a SQLite file ``mlruns.db``)

Scenario 3: MLflow on localhost with Tracking Server
----------------------------------------------------

Similar to scenario 1 but a tracking server is launched, listening for REST request calls at the default port 5000.
The arguments supplied to the ``mlflow server <args>`` dictate what backend and artifact stores are used.
The default is local `FileStore`. For example, if a user launched a tracking server as
``mlflow server --backend-store-uri sqlite:///mydb.sqlite``, then SQLite would be used for backend storage instead.

As in scenario 1, MLflow uses a local `mlruns` filesystem directory as a backend store and artifact store. With a tracking
server running, the MLflow client interacts with the tracking server via REST requests, as shown in the diagram.

.. figure:: _static/images/scenario_3.png

To store all runs' MLflow entities, the MLflow client interacts with the tracking server via a series of REST requests:

 * **Part 1a and b**:

  * The MLflow client creates an instance of a `RestStore` and sends REST API requests to log MLflow entities

>   * The Tracking Server creates an instance of a `FileStore` to save MLflow entities and writes directly to the local `mlruns` directory.

Reviewer: Nit: 
```suggestion
  * The Tracking Server creates an instance of a `FileStore` to save MLflow entities and writes directly to the local `mlruns` directory
```

Reviewer: Nevermind

reside on remote hosts. This example scenario depicts an architecture with a remote MLflow Tracking Server,
a Postgres database for backend entity storage, and an S3 bucket for artifact storage.

.. figure:: _static/images/scenario_4.png

To record all runs' MLflow entities, the MLflow client interacts with the tracking server via a series of REST requests:

 * **Part 1a and b**:

  * The MLflow client creates an instance of a `RestStore` and sends REST API requests to log MLflow entities
  * The Tracking Server creates an instance of an `SQLAlchemyStore` and connects to the remote host to
    insert MLflow entities in the database.

For artifact logging, the MLflow client interacts with the remote Tracking Server and artifact storage host:

 * **Part 2a, b, and c**:

  * The MLflow client uses `RestStore` to send a REST request to fetch the artifact store URI location from the Tracking Server
  * The Tracking Server responds with an artifact store URI location (an S3 storage URI in this case)
  * The MLflow client creates an instance of an `S3ArtifactRepository`, connects to the remote AWS host using the

>     `boto` client libraries, and uploads the artifacts to the S3 bucket URI location.

Reviewer: Can we also link to boto? can't hurt.

Author: done!

  `[Azure] <http://docs.microsoft.com/azure/databricks/applications/mlflow/access-hosted-tracking-server>`_, or :ref:`the quickstart <quickstart_logging_to_remote_server>` to
  easily get started with hosted MLflow on Databricks Community Edition.

How Runs and Artifacts are Recorded
===================================

> As mentioned above, MLflow runs can be recorded to local files, to a SQLAlchemy compatible database, or remotely to a tracking server.

Reviewer: Can we also restate that MLflow artifacts can be persisted to local files and a variety of remote file storage solutions?

Author: Added.

For storing runs and artifacts, MLflow uses two components for `storage <https://mlflow.org/docs/latest/tracking.html#storage>`_: backend store and artifact store. While the backend store persists
MLflow entities (runs, parameters, metrics, tags, notes, metadata, etc), the artifact store persists artifacts
(files, models, images, in-memory objects, or model summary, etc).

Which `backend store <https://mlflow.org/docs/latest/tracking.html#backend-stores>`_ or `artifact store <https://mlflow.org/docs/latest/tracking.html#artifact-stores>`_ is used by the MLFlow client or tracking server

> depends on four different scenarios.

Reviewer: There may be more than 4 scenarios here - the ones you're covering are representative of many common cases. Perhaps it's better to say something like "The MLflow client can interface with a variety of backend and artifact storage configurations. Here are four common configuration scenarios:"

Author: I like that!

> Scenario 1: MLflow on the localhost

Reviewer: Nit: let's drop the `the` from `the localhost` across these scenario titles, it's cleaner.

![image](https://user-images.githubusercontent.com/39497902/105538288-08648a80-5ca8-11eb-8e45-2b00ef8c5473.png)

```suggestion
Scenario 1: MLflow on localhost
```

Author: lol!

Author: done!

-----------------------------------

> A common scenario among developers who want to develop locally, MLflow runs on a localhost or a laptop, where both the

Reviewer: ```suggestion
Many developers run MLflow on their local machine, where both the
```

Author: done!

> backend store and artifact store is a directory on the local filesystem ``./mlruns,`` as shown in the diagram.

Reviewer: ```suggestion
backend store and artifact store share a directory on the local filesystem - ``./mlruns`` - as shown in the diagram.
```

Author: done!

The `MLflowClient` directly interfaces with an instance of a `FileStore` and `LocalArtifactRepository`.

.. figure:: _static/images/scenario_1.png

> In this simple scenario, the `MLFlowClient` is responsible to create instances of respective stores and
> interfaces with them to record MLflow entities and artifacts:

Reviewer: ```suggestion
In this simple scenario, the MLflow client uses the following interfaces to record MLflow entities and artifacts:
```

>  * `MLflowClient` creates an instance of a `LocalArtifactRepository` (to store artifacts)
>  * `MLflowClient` creates an instance of a `FileStore` (to save MLflow entities)

Reviewer: Small suggestion - also, can we link to the Python API docs for these?

```suggestion
 * An instance of a `LocalArtifactRepository` (to store artifacts)
 * An instance of a `FileStore` (to save MLflow entities)
```

Author: @dbczumar where do you want link to Python docs? In the first instance or occurrence of `MLflow client` here, since we are replacing or removing `MLflowClient` usage.

Reviewer: @dmatrix I'd like to link to `LocalArtifactRepository` and `FileStore` if possible (we can link to the GitHub source files using permalinks, since I realize now that we don't have docs for these on mlflow.org)

Scenario 2: MLflow on the localhost with SQLite
-----------------------------------------------

> A slight variation from scenario 1, MLflow runs on a localhost or laptop but with backend store as an SQLAlchemy compatible

Reviewer: ```suggestion
Many users also run MLflow on their local machines with a SQLAlchemy-compatible
```

Author: done!

> database type: **SQLite**. Artifacts are stored under the local ``./mlruns`` directory, and MLflow entities are

Reviewer: Small suggestion. Can we also link to the sqlite docs?

```suggestion
database: **SQLite**. In this case, artifacts are stored under the local ``./mlruns`` directory, and MLflow entities are
```

Author: done!

> inserted in an SQLite database file ``mlruns.db``.

Reviewer: ```suggestion
inserted in a SQLite database file ``mlruns.db``.
```

Author: done!

.. figure:: _static/images/scenario_2.png

> As above, `MLflowClient` is responsible to create respective instances:

Reviewer: In this scenario, the MLflow client uses the following interfaces to record MLflow entities and artifacts:

Author: done!

>  * `MLflowClient` creates an instance of a `LocalArtifactRepository` (to save artifacts)
>  * `MLflowClient` creates an instance of an `SQLAlchemyStore` (to save MLflow entities) and writes to SQLite file ``mlruns.db``)

Reviewer: ```suggestion
 * An instance of a `LocalArtifactRepository` (to save artifacts)
 * An instance of an `SQLAlchemyStore` (to save MLflow entities to an ``mlruns.db`` SQLite file)
```

Author: done!

Scenario 3: MLflow on the localhost with Tracking Server
--------------------------------------------------------

> Similar to scenario 1 but a tracking server is launched, listening for REST request calls at the default port 5000.

Reviewer: Can we indicate that this setup where the FileStore is used comes as a result of the user running `mlflow server` with default arguments? (If the user ran something like `mlflow server --backend-store-uri sqlite:///mydb.sqlite`, then SQL would be used for backend storage instead).

Author: I don't see why not by adding this further to specify which store is intended, based on either default or specified arguments to `mlflow server ....`.

Reviewer: Sounds good! Let's do that! (i.e. specify `--backend-store-uri mlruns`)

Reviewer: Cool! Thanks!

> As in scenario 1, MLflow uses the local `file://path/mlruns` as a backend store and artifact store. With a tracking

Reviewer: ```suggestion
As in scenario 1, MLflow uses a local `mlruns` filesystem directory as a backend store and artifact store. With a tracking
```

Author: done!

> server running, the `MLflowClient` will interact with tracking server via REST requests as shown in the diagram.

Reviewer: ```suggestion
server running, the MLflow client interacts with the tracking server via REST requests, as shown in the diagram.
```

.. figure:: _static/images/scenario_3.png

To store all runs' MLflow entities, the `MLflowClient` interacts with the tracking server via a series of REST requests:

 * **Part 1a and b**:

>   * `MLflowClient` creates an instance of a `RestStore`
>   * `MLflowClient` sends a REST API request to log MLflow entities

Reviewer: ```suggestion
  * The MLflow client creates an instance of a `RestStore` and sends REST API requests to log MLflow entities
```

>   * Tracking Server creates an instance of a `FileStore` (to save MLflow entities) and writes directly
>     to the local `file://path/mlruns` directory.

Reviewer: ```suggestion
  * The Tracking Server creates an instance of a `FileStore` (to save MLflow entities) and writes directly
    to the local `mlruns` directory.
```

Author: done!

For the artifacts, the `MLflowClient` interacts with the tracking server via a REST request:

 * **Part 2a, b, and c**:

>    * `MLflowClient` uses `RestStore` to send a REST request to fetch the artifact store URI location
>    * Tracking Server responds with an artifact store URI location
>    * `MLflowClient` creates an instance of a `LocalArtifactRepository` (to save artifacts at the local URI location)

Reviewer: ```suggestion
   * The MLflow client uses `RestStore` to send a REST request to fetch the artifact store URI location
   * The Tracking Server responds with an artifact store URI location
   * The MLflow client creates an instance of a `LocalArtifactRepository` and saves artifacts to the local filesystem location specified by the artifact store URI (a subdirectory of `mlruns`)
```

Author: done!

Scenario 4: MLflow with remote Tracking Server, backend and artifact stores
---------------------------------------------------------------------------

> This is a common MLflow setup where the tracking server is remote, to which clients
> send REST requests; a remote SQLAlchemy compatible `database-backend store <https://mlflow.org/docs/latest/tracking.html#backend-stores>`_, to which the tracking
> server records MLflow entities; and a remote, supported and scalable `artifact store <https://mlflow.org/docs/latest/tracking.html#artifact-stores>`_, to which the clients
> upload artifacts.

Reviewer: I think it would help to describe the general capability of MLflow to work in distributed environments and indicate that this is an example.

E.g.

```suggestion
MLflow also supports distributed architectures, where the Tracking Server, backend store, and artifact store reside on remote hosts. This example scenario depicts an architecture with a remote MLflow Tracking Server, a Postgres database for backend entity storage, and an S3 bucket for artifact storage.
```

Author: You don't like my parallel structure of using infinitive phrases ... :-) I Will update it with this distributed scenario.

Author: @dbczumar Thanks for detailed feedback. I have incorporated all nits and suggestions, including much crispier diagrams—the ppt did the trick.

.. figure:: _static/images/scenario_4.png

To record all runs' MLflow entities, the `MLflowClient` interacts with the tracking server via a series of REST requests:

 * **Part 1a and b**:

>   * `MLflowClient` creates an instance of a `RestStore`
>   * `MLflowClient` sends a REST API request to log MLflow entities

Reviewer: ```suggestion
  * The MLflow client creates an instance of a `RestStore` and sends REST API requests to log MLflow entities
```

Author: done!

  * The Tracking Server creates an instance of an `SQLAlchemyStore` and connects to the remote host to
    insert MLflow entities in the database.

> For the artifacts, the `MLflowClient` first interacts with the tracking server via a REST request:

Reviewer: ```suggestion
For artifact logging, the MLflow client interacts with the remote Tracking Server and artifact storage host:
```

Author: done!

 * **Part 2a, b, and c**:

>   * `MLflowClient` uses `RestStore` to send a REST request to fetch the artifact store URI location
>   * Tracking Server responds with an artifact store URI location, the name of the S3 bucket

Reviewer: ```suggestion
  * The MLflow client uses `RestStore` to send a REST request to fetch the artifact store URI location from the Tracking Server
  * The Tracking Server responds with an artifact store URI location (an S3 storage URI in this case)
```

Author: done!

>   * `MLflowClient` creates an instance of an `S3ArtifactRepository`, connects to the remote AWS host using the

Reviewer: ```suggestion
  * The MLflow client creates an instance of an `S3ArtifactRepository`, connects to the remote AWS host using the
```

Author: Done!

    `boto` client libraries, and uploads the artifacts to the S3 bucket URI location.

.. note::

>     In all scenarios, the MLflow client directly pushes artifacts to the artifact store. It does not proxy these through the
>     tracking server.

Reviewer: ```suggestion
    In all scenarios, the MLflow client directly logs artifacts to the remote artifact store. It does not proxy these through the
    tracking server.
```

Author: done!
